File: EIDSS v5/android.java/workspace/EIDSS/src/com/bv/eidss/LoadReferencesActivity.java
package com.bv.eidss;

import com.WSParser.WebServices.EidssService.VectorBaseReferenceRaw;
import com.WSParser.WebServices.EidssService.VectorBaseReferenceTranslationRaw;
import com.WSParser.WebServices.EidssService.VectorGisBaseReferenceRaw;
import com.WSParser.WebServices.EidssService.VectorGisBaseReferenceTranslationRaw;
import com.bv.eidss.data.EidssDatabase;
import com.bv.eidss.model.Options;
import com.bv.eidss.oa.OaClient;
import android.app.Activity;
import android.app.Dialog;
import android.app.ProgressDialog;
import android.os.AsyncTask;
import android.os.Bundle;
import android.view.View;
import android.widget.EditText;

public class LoadReferencesActivity extends Activity {
    private final int PROGRESS_DIALOG_ID = 0;
    public ProgressDialog mDialog;
    public int iErr = 0;
	
	private LoadReferencesActivity _this;
	public LoadReferencesActivity() {
    	_this = this;
	}

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setTitle(R.string.TitleLoadReferences);
        setContentView(R.layout.load_references_layout);

        EidssDatabase db = new EidssDatabase(_this);
        String org = db.Options().getLoginOrganization();
        String usr = db.Options().getLoginUsername();
        ((EditText)findViewById(R.id.LoginOrganization)).setText(org);
        ((EditText)findViewById(R.id.LoginName)).setText(usr);
        db.close();
        
    	findViewById(R.id.OkButton).setOnClickListener(new View.OnClickListener() {
			@SuppressWarnings("deprecation")
			public void onClick(View v) {
				_this.showDialog(PROGRESS_DIALOG_ID);
				new LoadReferenceTask().execute(_this);
			}
		});
    	findViewById(R.id.CancelButton).setOnClickListener(new View.OnClickListener() {
			public void onClick(View v) {
				_this.finish();
			}
		});
    }

    @Override
    protected Dialog onCreateDialog(int id) {
    	switch(id)
    	{
    		case PROGRESS_DIALOG_ID:
    			mDialog = ProgressDialog.show(this, "", getResources().getString(R.string.WaitLoading), true);
    			return mDialog;
            case R.string.ConnectionFailed:
            case R.string.LoginFailed:
            case R.string.AccessFailed:
                return EidssAndroidHelpers.ErrorDialog(this, getResources().getString(id));
    	}
    	return null;
    }
}


final class LoadReferenceTask extends AsyncTask<LoadReferencesActivity, Void, Void>{
	@Override
	protected Void doInBackground(LoadReferencesActivity... params) {
		final LoadReferencesActivity activity = params[0];
		activity.iErr = 0;
		VectorBaseReferenceRaw refs = null;
		VectorBaseReferenceTranslationRaw refs_trans = null; 
		VectorGisBaseReferenceRaw gis_refs = null;
		VectorGisBaseReferenceTranslationRaw gis_refs_trans = null;
		String org = "";
		String usr = "";
		try {
	        EidssDatabase db = new EidssDatabase(activity);
	        String url = db.Options().getServerUrl();
	        db.close();
	        org = ((EditText)activity.findViewById(R.id.LoginOrganization)).getText().toString();
	        usr = ((EditText)activity.findViewById(R.id.LoginName)).getText().toString();
	        String pwd = ((EditText)activity.findViewById(R.id.LoginPassword)).getText().toString();
	        OaClient client = new OaClient(url, org, usr, pwd);
	        refs = client.LoadReferences();
	        refs_trans = client.LoadReferenceTranslations();
	        gis_refs = client.LoadGisReferences();
	        gis_refs_trans = client.LoadGisReferenceTranslations();
		}
		catch(ClassCastException e){
			if (e.getMessage().startsWith("org.ksoap2.SoapFault"))
				activity.iErr = R.string.AccessFailed;
			else if (e.getMessage().startsWith("org.ksoap2.serialization.NullSoapObject"))
				activity.iErr = R.string.LoginFailed;
		}
		catch(Exception e){
			activity.iErr = R.string.ConnectionFailed;
		}

		if (activity.iErr == 0){
	        EidssDatabase db = new EidssDatabase(activity);
	        if (db.LoadReferences(refs, refs_trans, gis_refs, gis_refs_trans)){
	        	Options opt = db.Options();
                opt.setLoginOrganization(org);
                opt.setLoginUsername(usr);
                db.OptionsUpdate(opt);
	        }
	        db.close();
			
			activity.mDialog.dismiss();
			activity.finish();
		}
		else{
			activity.mDialog.dismiss();
			activity.runOnUiThread(new Runnable() {
	            @SuppressWarnings("deprecation")
				@Override
	            public void run() {
	            	activity.showDialog(activity.iErr);
	            }
	        });
		}
		return null;
	}
}

File: EIDSS v5/android/EIDSS/LoadReferencesActivity.cs
using System;
using System.ServiceModel;
using System.Threading;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Web.Services.Protocols;
using Android.App;
using Android.Content;
using Android.OS;
using Android.Runtime;
using Android.Views;
using Android.Widget;
using bv.com.eidss;
using eidssdroid.Model;
using eidssdroid.model.Database;
using eidss.model.Enums;

namespace EIDSS
{
    [Activity]
    public class LoadReferencesActivity : Activity
    {
        private const int PROGRESS_DIALOG_ID = 0;
        private ProgressDialog mDialog;

        protected override void OnCreate(Bundle bundle)
        {
            base.OnCreate(bundle);

            SetTitle(Resource.String.TitleLoadReferences);
            SetContentView(Resource.Layout.LoadReferencesLayout);

            Button buttonOk = FindViewById<Button>(Resource.Id.OkButton);
            Button buttonCancel = FindViewById<Button>(Resource.Id.CancelButton);
            EditText org = FindViewById<EditText>(Resource.Id.LoginOrganization);
            EditText name = FindViewById<EditText>(Resource.Id.LoginName);
            EditText pwd = FindViewById<EditText>(Resource.Id.LoginPassword);
            using (EidssDatabase db = new EidssDatabase(this))
            {
                var opt = db.Options();
                org.Text = opt.strLoginOrganization;
                name.Text = opt.strLoginUsername;
            }

            buttonOk.Click += delegate
            {
                ShowDialog(PROGRESS_DIALOG_ID);
                ThreadPool.QueueUserWorkItem(c =>
                {
                    int iErr = 0;
                    using (EidssDatabase db = new EidssDatabase(this))
                    {
                        var opt = db.Options();
                        string url = opt.strServerUrl;
                        BaseReferenceRaw[] refs = null;
                        BaseReferenceTranslationRaw[] refs_trans = null;
                        GisBaseReferenceRaw[] gis_refs = null;
                        GisBaseReferenceTranslationRaw[] gis_refs_trans = null;
                        bool bRes = false;
                        try
                        {
                            OAClient oa = new OAClient(url);
                            bRes = oa.Proxy.LoadReferences(org.Text, name.Text, pwd.Text,
                                                                new ArrayOfLong()
                                                                {
                                                                    (long)BaseReferenceType.rftDiagnosis,
                                                                    (long)BaseReferenceType.rftHumanAgeType,
                                                                    (long)BaseReferenceType.rftHumanGender,
                                                                    (long)BaseReferenceType.rftFinalState,
                                                                    (long)BaseReferenceType.rftHospStatus
                                                                },
                                                                new ArrayOfString() { "en", "ru", "ka" },
                                                                out refs, out refs_trans, out gis_refs, out gis_refs_trans);
                        }
                        catch (EndpointNotFoundException) 
                        {
                            iErr = Resource.String.ConnectionFailed; // java.lang.IllegalArgumentException, org.xmlpull.v1.XmlPullParserException, java.net.SocketTimeoutException
                        }
                        catch (FaultException) 
                        {
                            iErr = Resource.String.AccessFailed; // java.lang.ClassCastException: org.ksoap2.SoapFault cannot be cast to org.ksoap2.serialization.SoapObject
                        }
                        catch (CommunicationException)
                        {
                            iErr = Resource.String.ConnectionFailed;
                        }
                        catch (Exception)
                        {
                            iErr = Resource.String.ConnectionFailed;
                        }
                        if (iErr == 0 && !bRes)
                        {
                            iErr = Resource.String.LoginFailed; // java.lang.ClassCastException: org.ksoap2.serialization.NullSoapObject cannot be cast to org.ksoap2.serialization.SoapObject
                        }
                        if (bRes)
                        {
                            bRes = db.LoadReferences(refs, refs_trans, gis_refs, gis_refs_trans);
                            if (bRes)
                            {
                                opt.strLoginOrganization = org.Text;
                                opt.strLoginUsername = name.Text;
                                db.OptionsUpdate(opt);

                                this.Finish();
                            }
                        }
                    }
                    mDialog.Dismiss();

                    if (iErr != 0)
                        RunOnUiThread(() => ShowDialog(iErr));
                });
            };

            buttonCancel.Click += delegate
            {
                this.Finish();
            };
        }

        protected override Dialog OnCreateDialog(int id)
        {
            switch (id)
            {
                case PROGRESS_DIALOG_ID:
                    mDialog = ProgressDialog.Show(this, "", Resources.GetString(Resource.String.WaitLoading), true);
                    return mDialog;
                case Resource.String.ConnectionFailed:
                case Resource.String.LoginFailed:
                case Resource.String.AccessFailed:
                    return EidssAndroidHelpers.ErrorDialog(this, Resources.GetString(id));
            }
            return null;
        }

    }
}